File: internal/scaffold/render.go
package scaffold

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"cup/internal/tmpl"
	"cup/internal/ui"
)

var placeholderRe = regexp.MustCompile(`\{\{[^}]+\}\}`)

// Render loads template <family>/<kind>/<name> (preferring the project's own
// copy under .cup/templates/<kind>/), substitutes every {{key}} from vars, and
// fails if any placeholder was left unresolved. Substitution runs to a fixed
// point, so a variable whose value itself contains a placeholder (e.g. a
// {{hello}} greeting embedding {{name}}) is fully resolved.
func Render(root, family, kind, name string, vars map[string]string) (string, error) {
	raw, err := tmpl.Read(root, family, kind, name)
	if err != nil {
		return "", fmt.Errorf("template %s/%s not found", kind, name)
	}
	content := string(raw)
	for i := 0; i <= len(vars); i++ {
		before := content
		for k, v := range vars {
			content = strings.ReplaceAll(content, "{{"+k+"}}", v)
		}
		if content == before {
			break
		}
	}
	if left := placeholderRe.FindAllString(content, -1); len(left) > 0 {
		set := map[string]bool{}
		for _, l := range left {
			set[l] = true
		}
		uniq := make([]string, 0, len(set))
		for l := range set {
			uniq = append(uniq, l)
		}
		sort.Strings(uniq)
		return "", fmt.Errorf("template %s/%s has unresolved placeholders: %s",
			kind, name, strings.Join(uniq, ", "))
	}
	return content, nil
}

// rel renders path relative to root for tidy log output.
func rel(root, path string) string {
	if r, err := filepath.Rel(root, path); err == nil {
		return r
	}
	return path
}

// WriteFile writes content to path, prompting before overwriting an existing
// file. It reports true if written, false if the user declined the overwrite —
// a legitimate skip, not an error.
func WriteFile(root, path, content string) (bool, error) {
	if _, err := os.Stat(path); err == nil {
		ok, err := ui.Confirm(fmt.Sprintf("%s exists. overwrite?", rel(root, path)), false)
		if err != nil {
			return false, err
		}
		if !ok {
			ui.Skipped(rel(root, path))
			return false, nil
		}
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return false, err
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		return false, err
	}
	ui.Wrote(rel(root, path))
	return true, nil
}

// EnsureFile creates path with content only if it does not already exist.
func EnsureFile(root, path, content string) error {
	if _, err := os.Stat(path); err == nil {
		return nil
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		return err
	}
	ui.Wrote(rel(root, path))
	return nil
}

File: cpp/src/libs/cup/scaffold/Render.cpp
module;
#include <expected>
#include <filesystem>
#include <format>
#include <fstream>
#include <ios>
#include <map>
#include <regex>
#include <set>
#include <string>
#include <string_view>
#include <system_error>
#include <utility>
#include <vector>
module cup.scaffold;

import cup.tmpl;
import cup.ui;

namespace cup::scaffold {
namespace {

std::string replace_all(std::string s, std::string_view from, std::string_view to) {
    if (from.empty()) {
        return s;
    }
    std::string out;
    out.reserve(s.size());
    std::size_t pos = 0;
    while (true) {
        const auto found = s.find(from, pos);
        if (found == std::string::npos) {
            out.append(s, pos, std::string::npos);
            break;
        }
        out.append(s, pos, found - pos);
        out.append(to);
        pos = found + from.size();
    }
    return out;
}

const std::regex& placeholder_regex() {
    static const std::regex re(R"(\{\{[^}]+\}\})");
    return re;
}

// Deduplicated, sorted (std::set orders by std::string's operator<, matching
// Go's sort.Strings) so the error message lists each unresolved placeholder
// once, in a stable order.
std::vector<std::string> unresolved_placeholders(const std::string& content) {
    std::set<std::string> unique;
    for (auto it = std::sregex_iterator(content.begin(), content.end(), placeholder_regex());
         it != std::sregex_iterator(); ++it) {
        unique.insert(it->str());
    }
    return {unique.begin(), unique.end()};
}

std::string join(const std::vector<std::string>& items, std::string_view sep) {
    std::string out;
    for (std::size_t i = 0; i < items.size(); ++i) {
        if (i != 0) {
            out += sep;
        }
        out += items[i];
    }
    return out;
}

std::expected<void, error::Error> create_dir(const std::filesystem::path& dir) {
    if (dir.empty()) {
        return {};
    }
    std::error_code ec;
    std::filesystem::create_directories(dir, ec);
    if (ec) {
        return std::unexpected(
            error::Error(std::format("creating {}: {}", dir.string(), ec.message())));
    }
    return {};
}

std::expected<void, error::Error> write_bytes(const std::filesystem::path& path,
                                              std::string_view content) {
    std::ofstream out(path, std::ios::binary | std::ios::trunc);
    out << content;
    if (!out) {
        return std::unexpected(error::Error(std::format("writing {}", path.string())));
    }
    return {};
}

}

std::expected<std::string, error::Error> render(const std::filesystem::path& root,
                                                 std::string_view family, std::string_view kind,
                                                 std::string_view name,
                                                 const std::map<std::string, std::string>& vars) {
    auto raw = tmpl::read(root, family, kind, name);
    if (!raw.has_value()) {
        return std::unexpected(error::Error(std::format("template {}/{} not found", kind, name)));
    }
    std::string content = std::move(*raw);
    for (std::size_t i = 0; i <= vars.size(); ++i) {
        const std::string before = content;
        for (const auto& [k, v] : vars) {
            content = replace_all(std::move(content), "{{" + k + "}}", v);
        }
        if (content == before) {
            break;
        }
    }
    if (auto left = unresolved_placeholders(content); !left.empty()) {
        return std::unexpected(error::Error(std::format(
            "template {}/{} has unresolved placeholders: {}", kind, name, join(left, ", "))));
    }
    return content;
}

std::string rel(const std::filesystem::path& root, const std::filesystem::path& path) {
    const auto relative = path.lexically_relative(root);
    if (relative.empty()) {
        return path.string();
    }
    return relative.string();
}

std::expected<bool, error::Error> write_file(const std::filesystem::path& root,
                                             const std::filesystem::path& path,
                                             std::string_view content) {
    std::error_code ec;
    if (std::filesystem::exists(path, ec)) {
        auto confirmed = ui::confirm(std::format("{} exists. overwrite?", rel(root, path)), false);
        if (!confirmed.has_value()) {
            return std::unexpected(std::move(confirmed).error());
        }
        if (!*confirmed) {
            ui::skipped(rel(root, path));
            return false;
        }
    }
    if (auto created = create_dir(path.parent_path()); !created.has_value()) {
        return std::unexpected(std::move(created).error());
    }
    if (auto written = write_bytes(path, content); !written.has_value()) {
        return std::unexpected(std::move(written).error());
    }
    ui::wrote(rel(root, path));
    return true;
}

std::expected<void, error::Error> ensure_file(const std::filesystem::path& root,
                                              const std::filesystem::path& path,
                                              std::string_view content) {
    std::error_code ec;
    if (std::filesystem::exists(path, ec)) {
        return {};
    }
    if (auto created = create_dir(path.parent_path()); !created.has_value()) {
        return created;
    }
    if (auto written = write_bytes(path, content); !written.has_value()) {
        return written;
    }
    ui::wrote(rel(root, path));
    return {};
}

}
